flink-formats/flink-avro-apicurio-registry/src/main/java/org/apache/flink/formats/avro/registry/apicurio/ApicurioSchemaRegistryCoder.java

```diff
@@ -0,0 +1,379 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.formats.avro.registry.apicurio;
+
+import org.apache.flink.formats.avro.AvroFormatOptions;
+import org.apache.flink.formats.avro.SchemaCoder;
+
+import io.apicurio.registry.rest.client.RegistryClient;
+import io.apicurio.registry.rest.v2.beans.ArtifactMetaData;
+import io.apicurio.registry.rest.v2.beans.ArtifactReference;
+import io.apicurio.registry.rest.v2.beans.IfExists;
+import io.apicurio.registry.rest.v2.beans.VersionMetaData;
+import io.apicurio.registry.types.ArtifactType;
+import org.apache.avro.Schema;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.formats.avro.registry.apicurio.AvroApicurioFormatOptions.USE_GLOBALID;
+import static org.apache.flink.formats.avro.registry.apicurio.AvroApicurioFormatOptions.USE_HEADERS;
+
+/** Reads and Writes schema using Avro Schema Registry protocol. */
+public class ApicurioSchemaRegistryCoder implements SchemaCoder {
+
+    public static final String APICURIO_KEY_GLOBAL_ID_HEADER = "apicurio.key.globalId";
+
+    public static final String APICURIO_KEY_CONTENT_ID_HEADER = "apicurio.key.contentId";
+
+    public static final String APICURIO_VALUE_GLOBAL_ID_HEADER = "apicurio.value.globalId";
+
+    public static final String APICURIO_VALUE_CONTENT_ID_HEADER = "apicurio.value.contentId";
+
+    public static final String APICURIO_VALUE_ENCODING = "apicurio.value.encoding";
+
+    private final RegistryClient registryClient;
+
+    private final Map<String, Object> configs;
+
+    private static final int MAGIC_BYTE = 0;
+
+    public static final String TOPIC_NAME = "TOPIC_NAME";
+
+    public static final String HEADERS = "HEADERS";
+
+    public static final String IS_KEY = "IS_KEY";
+
+    private static final Logger LOG = LoggerFactory.getLogger(ApicurioSchemaRegistryCoder.class);
+
+    /**
+     * Creates {@link SchemaCoder} that uses provided {@link RegistryClient} to connect to schema
+     * registry.
+     *
+     * @param registryClient client to connect schema registry
+     * @param configs map for registry configs
+     */
+    public ApicurioSchemaRegistryCoder(RegistryClient registryClient, Map<String, Object> configs) {
+        this.registryClient = registryClient;
+        this.configs = configs;
+    }
+
+    @Override
+    public Schema readSchema(InputStream in) throws IOException {
+        throw new IOException(
+                "Incompatible Kafka connector. Use a Kafka connector that passes headers");
```

Comment:
What's the connection between `readSchema()` and 'a Kafka connector that passes headers' ?